Persistent VM documentation (#429)
netgroup-polito/CrownLabs: operators/README.md

Based on [Kubebuilder 2.3](https://github.com/kubernetes-sigs/kubebuilder.git), the operator implements the environment creation logic of CrownLabs.

> In the following, a little scheme useful to understand the general architecture of the Instance Operator.

Reviewer: ```suggestion
The next picture shows a general architecture of the Instance Operator.
```

> On the left you can see the controller (in blue) and the two CRD used to describe the desired status (in green). On the right it is possible to view the environment that is created. It is divided in two parts, one is the set of components that are created only in persistent case (that will be analyzed in _Persistent Feature_ section), and the other one is the set of resources created even in not persistent case (the VirtualMachine Instance itself and the components useful to connect to it, e.g. secret, service and ingress).

Reviewer 2: ```suggestion
On the left you can see the controller (in blue) and the two CRDs used to describe the desired status (in green) and better detailed below. On the right it is possible to view the set of resources that are created for each environment. It is divided into two parts, one representing the components that are created only in the persistent case (that will be analyzed in the _Persistent Feature_ section), while the other one is the set of resources created in both scenarios (i.e. the VirtualMachine Instance itself and the components useful to connect to it, e.g. secret, service and ingress).
```

> On the left you can see the controller (in blue) and the two CRD used to describe the desired status (in green). On the right it is possible to view the environment that is created. It is divided in two parts, one is the set of components that are created only in persistent case (that will be analyzed in _Persistent Feature_ section), and the other one is the set of resources created even in not persistent case (the VirtualMachine Instance itself and the components useful to connect to it, e.g. secret, service and ingress).

Reviewer: ```suggestion
On the left you can see the controller (in blue) and the two CRDs used to describe the desired status (in green).
The right side shows the environment that is created.
A first part depicts the set of components that are created only in _persistent_ case (that will be analyzed in _Persistent Feature_ section).
A second part shows the set of resources created even in not persistent case (the VirtualMachine Instance itself and the components needed to connect to it, e.g. secret, service and ingress).
```

![Instance Operator Architecture](../documentation/instance-operator.svg)

Upon the creation of a *Instance*, the operator triggers the creation of the following components:
* Kubevirt VirtualMachine Instance and the logic to access the noVNC instance inside the VM (Service, Ingress)
* An instance of [Oauth2 Proxy](https://github.com/oauth2-proxy/oauth2-proxy) (Deployment, Service, Ingress) to regulate access to the VM.

> *  A [DataVolume](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/datavolumes.md) (only in case of prersistent VMs). It wraps a Persistent Volume Claim (PVC), and takes care of initializing it with the content of the selected VM image through an importer pod.

Reviewer 2: ```suggestion
*  A [DataVolume](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/datavolumes.md) (only in case of persistent VMs). It wraps a Persistent Volume Claim (PVC), and takes care of initializing it with the content of the selected VM image through an importer pod.
```

Based on [Kubebuilder 2.3](https://github.com/kubernetes-sigs/kubebuilder.git), the operator implements the environment creation logic of CrownLabs.

> ![Instance Operator Architecture](../documentation/instance-operator.jpg)

Reviewer 2: Three comments here:
1. From my point of view, the image should be accompanied by some descriptive text to contextualize what the figure represents and which are the main elements to look at. Otherwise it may be difficult for a newbie reader to understand many concepts. 
2. There are some inaccuracies. For instance, the VM is also created by the operator, the oauth2-proxy components are missing, and it is not clear to me the meaning of the arrow going from the secret/service/ingress to the VMI.
3. It would be better if the picture were in the svg format, as it scales better and it prevents the issues typical of raster graphics. Additionally, it could be useful to check-in also the source file, to allow for easy modifications if the workflow is modified (e.g. docker-based instances).

In order to use this feature, [the CDI operator](../infrastructure/virtualization/README.md) has to be installed.

Granted the CDI Operator has been deployed, the Instance Operator spawns a persistent VM similarly to a normal (i.e. non-persistent) one. To this end, the selection about which kind of VM is wanted, i.e. persistent or not, is performed looking at the corresponding field in the `template` spec. In the same way you can choose the size of the persistent disk that the VM will have.

We have decided to use a VM instead of directly create a VirtualMachine Instance (VMI) because of the start and stop process already managed by KubeVirt. In this way the VMI is easily stopped and started again in the user interface (what happens is that the running field in the Instance resource is modified).

It is important to notice that the Instance Operator has to watch not only to the Instance but also the DataVolume associated to it, that is to say that the controller has to monitor these two resources, so that in case one of them is changed the controller is triggered and start a reconcile in order to achieve a consistent state in the cluster.

> Since we are interested only in one specific change in the DataVolume (the status Phase becoming Succeeded) and lot of changes in the Instance do not need controller's actions (those in the status) a [Predicate](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) has been used. That fuction act before the reconcile, working as a filter and letting the reconcile function be called only in the case explained before.

Author: This link can be changed to the Crownlabs site's one when our PR will be merged

### Persistent Feature
The Instance Operator allows you to create persistent Virtual Machines. Persistent means that the VM can be stopped and restarted, deleted and recreated without data loss. To make this possible, the content of the VM's disk is stored in a PVC.
In order to use this feature, [the CDI operator](../infrastructure/virtualization/README.md) has to be installed.

> Granted the CDI Operator has been deployed, the Instance Operator spawns a persistent VM similarly to a normal (i.e. non-persistent) one. To this end, the selection about which version to create is performed looking at the corresponding field in the `template` spec. It is also possible to select the amount of memory needed for the VM by modifying the corresponding `template` spec.

Reviewer 2: The last sentence is not clear to me. First, I suppose you are talking about disk size, not memory. Second, I would say that the disk is created of the size specified by the template (IMO the word *modifying* is misleading, as it is not clear that it is configured by the manager of the template and not by the operator).

> It is important to notice that the Instance Operator has to watch not only to the instance but also the datatvolume associated to it.

Reviewer 2: Please, be more explicit. What does watch mean? How is that configured?

> Once the instance is created, the controller starts creating a datavolume, whose creation triggers the CDI controller which creates two resources: the PVC and the importer pod.

Reviewer 2: ```suggestion
Once an Instance targeting a persistent VM is created, the operator starts creating a DataVolume resource. This, in turn, triggers the CDI controller which creates two resources: the PVC and the importer pod.
```

Please, explain the purpose of the PVC

> The aim of the importer pod is to exctract an image from the docker registry and load it inside the PVC. This process, depending on the size of the image, can take some minutes. Once the import is completed the Instance Operator wakes up (this is the only one case in which a datavolume resource is able to to trigger a reconcile, thanks to the filter and predicates used) and all other resources are created: service, secret, ingress, oauth2 ones and VM (and VMI in case the instance running field is set to true).

Reviewer 2: ```suggestion
The aim of the importer pod is to extract an image from the docker registry and load it inside the PVC. This process, depending on the size of the image, can take some minutes. Once the import is completed, the Instance Operator wakes up (this is the only case in which a DataVolume resource modification triggers a reconcile, thanks to the filter and predicates used) and all the other resources are created as described in the following.
```

Please:
* Double check for typos and be consistent with the rest of the text;
* Introduce concepts that may not be clear (I doubt a newbie reader would know what you mean by filter and predicate, you should at least add a link to the code and then describe a little the approach). Also the difference between VirtualMachine and VirtualMachineInstance deserves to be explained more in detail, as it is a key aspect of the design. Which docker registry are you referring to?
* Do not use abbreviations if not previously introduced.

> N.B. The process of creating a persistent VirtualMachine can take, as said, a bit more time with the respect to a normal one (5-10 mins). However this time is only needed for the creation of the PVC (only the first time the VM is created), so if VM is stopped, deleted (without deleting PVC) the recreation process will be way more fast.

Reviewer 2: I feel the previous alternative was clearer. I would just change *recreate* with *stop and restart*.

- **Instance** defines an instance of a certain template. The manipulation of those objects triggers the reconciliation logic in the operator, which creates/destroy associated resources (e.g.; Virtual Machines).

### Persistent Feature

> The Instance Operator allows you to create persistent VirtualMachines (VM). Persistent means that the VM can be stopped and restarted, deleted and recreated without data loss. 

Reviewer 2: ```suggestion
The Instance Operator allows you to create persistent Virtual Machines (VM). Persistent means that the VM can be stopped and restarted, deleted and recreated without data loss. 
```

In order to use this feature, [the CDI operator](../infrastructure/virtualization/README.md) has to be installed.

Provided that the CDI Operator has been deployed, the Instance Operator spawns a persistent VM similarly to a normal (i.e. non-persistent) one. To this end, the selection about which kind of VM is wanted, i.e. persistent or not, is performed looking at the corresponding field in the `template` spec. In the same way you can choose the size of the persistent disk that the VM will have.

> It has been decided to use a VM instead of directly creating a VirtualMachine Instance (VMI) because the start and stop process is already managed by KubeVirt. In this way the VMI is easily stopped and started again modifying the running field of Instance resource (this can be easily done from the user interface).

Reviewer 2: ```suggestion
Focusing on the resources created by the controller, it is worth highlighting that in the persistent case the Instance Operator instantiates a KubeVirt VirtualMachine object instead of a simple VirtualMachineInstance one. Indeed, this design choice allows to delegate the start and stop process of a VM to KubeVirt itself. Hence, a persistent VM can be easily stopped and started again modifying the `running` field of the Instance resource (optionally through the graphical user interface).
```

> It is important to notice that the Instance Operator has to watch not only the Instance but also the DataVolume associated to it. In other words when an Instance or a DataVolume is modified the controller is triggered and it starts a reconcile in order to achieve a consistent state in the cluster. Since only one specific change in the DataVolume is important (the status Phase becoming Succeeded) and a lot of changes in the Instance do not need controller actions (those in the status) a  [Predicate](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) has been used. That function acts before the reconcile, working as a filter and letting the reconcile function be called only in the case explained before.

Reviewer 2: ```suggestion
It is important to notice that the Instance Operator has to watch not only Instance resources but also the DataVolumes associated to them. In other words when an Instance or a DataVolume is modified, the controller is triggered and it starts a reconcile process in order to achieve a consistent state in the cluster. Since only one specific change in the DataVolume is important (i.e., when the status Phase becoming Succeeded) and a lot of changes in the Instance do not need controller actions (e.g., those affecting the status) a [Predicate](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) has been used. That function acts before the reconcile, working as a filter and letting the reconcile function be called only in the cases explained before.
```

> Once the Instance is deployed, the controller starts creating a DataVolume, which triggers the CDI controller that generates two resources: the PVC and the importer pod. 

Reviewer 2: ```suggestion
Focusing now, on the operator workflow, once the Instance is deployed, the controller starts creating a DataVolume, which triggers the CDI controller that generates two resources: the PVC and the importer pod. 
```

> It is important to notice that the Instance Operator has to watch not only Instance resources but also the DataVolumes associated to them. In other words when an Instance or a DataVolume is modified, the controller is triggered and it starts a reconcile process in order to achieve a consistent state in the cluster. Since only one specific change in the DataVolume is important (i.e., when the status Phase becoming Succeeded) and a lot of changes in the Instance do not need controller actions (e.g., those affecting the status) a [Predicate](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) has been used. That function acts before the reconcile, working as a filter and letting the reconcile function be called only in the cases explained before.

Reviewer 2: ```suggestion
It is important to notice that the Instance Operator has to watch not only Instance resources but also the DataVolumes associated to them. In other words when an Instance or a DataVolume is modified, the controller is triggered and it starts a reconcile process in order to achieve a consistent state in the cluster. Since only one specific change in the DataVolume is important (i.e., when the status Phase becomes Succeeded) and a lot of changes in the Instance do not need controller actions (e.g., those affecting the status) a [Predicate](https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/) has been used. That function acts before the reconcile, working as a filter and letting the reconcile function be called only in the cases explained before.
```

> Focusing now, on the operator workflow, once the Instance is deployed, the controller starts creating a DataVolume, which triggers the CDI controller that generates two resources: the PVC and the importer pod. 

Reviewer 2: ```suggestion
Focusing now on the operator workflow, once the Instance is deployed, the controller starts creating a DataVolume, which triggers the CDI controller that generates two resources: the PVC and the importer pod. 
```
